File: beintoosdk/mobile/android/src/com/beintoo/beintoosdk/BeintooVgood.java
/*******************************************************************************
 * Copyright 2011 Beintoo
 * 
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *   http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 ******************************************************************************/
package com.beintoo.beintoosdk;

import com.beintoo.beintoosdkutility.BeintooSdkParams;
import com.beintoo.beintoosdkutility.HeaderParams;

import com.beintoo.main.Beintoo;
import com.beintoo.wrappers.Message;
import com.beintoo.wrappers.Vgood;
import com.beintoo.wrappers.VgoodChooseOne;
import com.google.beintoogson.Gson;

public class BeintooVgood {
	String apiPreUrl = null;
	
	public BeintooVgood() {
		if(!BeintooSdkParams.useSandbox)
			apiPreUrl = BeintooSdkParams.apiUrl;
		else
			apiPreUrl = BeintooSdkParams.sandboxUrl;
	}

	/**
	 * Retrieve a vgood for the requested user
	 * 
	 * @param guid user guid
	 * @param codeID (optional) a string that represents the position in your code. We will use it to indentify different api calls of the same nature.
	 * @param latitude
	 * @param longitude
	 * @param radius
	 * @param privategood a bool to set if the vgood has to be private
	 * @return a json object with vgood data
	 */
	public Vgood getVgood(String guid, String codeID,String latitude, String longitude, String radius, boolean privategood) {
		
		String apiUrl = apiPreUrl+"vgood/get/byguid/"+guid+"/?allowBanner=true";
		
		if(privategood == true) apiUrl = apiUrl + "&private=true"; else apiUrl = apiUrl + "&private=false";  
		if(latitude != null) apiUrl = apiUrl + "&latitude="+latitude;
		if(longitude != null) apiUrl = apiUrl + "&longitude="+longitude;
		if(radius != null) apiUrl = apiUrl + "&radius="+radius;
		
		HeaderParams header = new HeaderParams();
		header.getKey().add("apikey");
		header.getValue().add(DeveloperConfiguration.apiKey);
		header.getKey().add("guid");
		header.getValue().add(guid);
		if(codeID != null){
			header.getKey().add("codeID");
			header.getValue().add(codeID);
		}
		
		try {
		// ADD THE USER AGENT
			String userAgent = Beintoo.userAgent;
			if(userAgent != null){
				header.getKey().add("User-Agent");
				header.getValue().add(userAgent);
			}
		}catch (Exception e){e.printStackTrace();}
		
		BeintooConnection conn = new BeintooConnection();
		String json = conn.httpRequest(apiUrl, header, null);
		Gson gson = new Gson();
		Vgood vgood = new Vgood();
		vgood = gson.fromJson(json, Vgood.class);
		
		return vgood;
	}
	
	/**
	 * Retrieve a vgood for the requested user
	 * 
	 * @param guid user guid
	 * @param codeID (optional) a string that represents the position in your code. We will use it to indentify different api calls of the same nature.
	 * @param latitude
	 * @param longitude
	 * @param radius
	 * @param privategood a bool to set if the vgood has to be private
	 * @return a json object with vgood data
	 */
	public VgoodChooseOne getVgoodList(String guid, String codeID,String latitude, String longitude, String radius, boolean privategood) {
		
		String apiUrl = apiPreUrl+"vgood/byguid/"+guid+"/?allowBanner=true";
		
		if(privategood == true) apiUrl = apiUrl + "&private=true"; else apiUrl = apiUrl + "&private=false";  
		if(latitude != null) apiUrl = apiUrl + "&latitude="+latitude;
		if(longitude != null) apiUrl = apiUrl + "&longitude="+longitude;
		if(radius != null) apiUrl = apiUrl + "&radius="+radius;
		
		HeaderParams header = new HeaderParams();
		header.getKey().add("apikey");
		header.getValue().add(DeveloperConfiguration.apiKey);
		header.getKey().add("guid");
		header.getValue().add(guid);
		if(codeID != null){
			header.getKey().add("codeID");
			header.getValue().add(codeID);
		}
		
		try {
		// ADD THE USER AGENT
			String userAgent = Beintoo.userAgent;
			if(userAgent != null){
				header.getKey().add("User-Agent");
				header.getValue().add(userAgent);
			}
		}catch (Exception e){e.printStackTrace();}
		
		BeintooConnection conn = new BeintooConnection();
		String json = conn.httpRequest(apiUrl, header, null);
		
		VgoodChooseOne vgoods = new Gson().fromJson(json, VgoodChooseOne.class);
		
		return vgoods;
	}

	
	/**
	 * Returns an array with all the vgoods earned by the user
	 * 
	 * @param userExt usr_ext unique identifier
	 * @param codeID (optional) a string that represents the position in your code. We will use it to indentify different api calls of the same nature.
	 * @return a Vgood object array
	 */
	public Vgood [] showByUser(String userExt, String codeID, boolean onlyConverted){
		String apiUrl = null;
		if(!onlyConverted)
			apiUrl = apiPreUrl+"vgood/show/byuser/"+userExt;
		else
			apiUrl = apiPreUrl+"vgood/show/byuser/"+userExt+"/?onlyConverted=true";
		
		HeaderParams header = new HeaderParams();
		header.getKey().add("apikey");
		header.getValue().add(DeveloperConfiguration.apiKey);
		
		if(codeID != null){
			header.getKey().add("codeID");
			header.getValue().add(codeID);
		}
		
		BeintooConnection conn = new BeintooConnection();
		String json = conn.httpRequest(apiUrl, header, null);
		Gson gson = new Gson();
		
		Vgood[] vgood = gson.fromJson(json, Vgood[].class);
		
		return vgood;		
	}
	
	/**
	 * Determinate if a player is eligible to receive a new virtual good
	 * 
	 * @param guid user guid
	 * @param codeID (optional) a string that represents the position in your code. We will use it to indentify different api calls of the same nature.
	 * @param latitude
	 * @param longitude
	 * @param radius
	 * @param privategood a bool to set if the vgood has to be private
	 * @return
	 */
	public Message isEligibleForVgood(String guid, String codeID,String latitude, String longitude, String radius) {
		
		String apiUrl = apiPreUrl+"vgood/iseligible/byguid/"+guid+"/?";
		
		if(latitude != null) apiUrl = apiUrl + "latitude="+latitude;
		if(longitude != null) apiUrl = apiUrl + "&longitude="+longitude;
		if(radius != null) apiUrl = apiUrl + "&radius="+radius;
		
		HeaderParams header = new HeaderParams();
		header.getKey().add("apikey");
		header.getValue().add(DeveloperConfiguration.apiKey);
		header.getKey().add("guid");
		header.getValue().add(guid);
		if(codeID != null){
			header.getKey().add("codeID");
			header.getValue().add(codeID);
		}
		
		BeintooConnection conn = new BeintooConnection();
		String json = conn.httpRequest(apiUrl, header, null);
		
		Message resp = new Gson().fromJson(json, Message.class);
		
		return resp;
	}
	
	/**
	 * Send a Vgood as a gift from a user to another user
	 * 
	 * @param vgoodExt the vgood id 
	 * @param userExt the user id of the gift sender
	 * @param userExt_to the user id of the gift sender
	 * @param codeID (optional) a string that represents the position in your code. We will use it to indentify different api calls of the same nature.
	 * @return a Message object with the status of the sendAsAgift operation
	 */
	public Message sendAsAGift (String vgoodExt, String userExt, String userExt_to, String codeID){
			
		String apiUrl = apiPreUrl+"vgood/sendasgift/"+vgoodExt+"/"+userExt+"/"+userExt_to;
	
		HeaderParams header = new HeaderParams();
		header.getKey().add("apikey");
		header.getValue().add(DeveloperConfiguration.apiKey);
		
		if(codeID != null){
			header.getKey().add("codeID");
			header.getValue().add(codeID);
		}
		
		BeintooConnection conn = new BeintooConnection();
		String json = conn.httpRequest(apiUrl, header, null);
		Gson gson = new Gson();
		
		Message msg = gson.fromJson(json, Message.class);
		
		return msg;	
	}

	/**
	 * Accept a vgood from a user
	 * 
	 * @param vgoodExt the vgood id 
	 * @param userExt the user id of the gift sender
	 * @param codeID (optional) a string that represents the position in your code. We will use it to indentify different api calls of the same nature.
	 * @return a Message object with the status of the sendAsAgift operation
	 */
	public Vgood acceptVgood (String vgoodExt, String userExt, String codeID){
			
		String apiUrl = apiPreUrl+"vgood/accept/"+vgoodExt+"/"+userExt;
	
		HeaderParams header = new HeaderParams();
		header.getKey().add("apikey");
		header.getValue().add(DeveloperConfiguration.apiKey);
		
		if(codeID != null){
			header.getKey().add("codeID");
			header.getValue().add(codeID);
		}
		
		BeintooConnection conn = new BeintooConnection();
		String json = conn.httpRequest(apiUrl, header, null);
		Gson gson = new Gson();
		
		Vgood v = gson.fromJson(json, Vgood.class);
		
		return v;	
	}
	
	/**
	 * Post a user vgood on the marketplace
	 * 
	 * @param vgoodExt the user id of the vgood
	 * @param userExt the user id of the user
	 * @return a message with the operation result
	 */
	public Message sellVgood (String vgoodExt, String userExt){
		String apiUrl = apiPreUrl+"vgood/marketplace/sell/"+vgoodExt+"/"+userExt;
		
		HeaderParams header = new HeaderParams();
		header.getKey().add("apikey");
		header.getValue().add(DeveloperConfiguration.apiKey);
		
		BeintooConnection conn = new BeintooConnection();
		String json = conn.httpRequest(apiUrl, header, null,true);
		Gson gson = new Gson();
		
		Message msg = gson.fromJson(json, Message.class);
		
		return msg;	
	}
	
	/**
	 * Return a list of vgoods on sale in the marketplace
	 * @param latitude the user latitude
	 * @param longitude the user longitude
	 * @param rows how many vgood you want to retrieve
	 * @param featured 
	 * @return a list of vgoods
	 */
	public Vgood[] showMarketPlace (String latitude, String longitude, int rows, boolean featured){
		String apiUrl = "";
		if(featured)
			apiUrl = apiPreUrl+"vgood/marketplace/show/?featured=true";
		else
			apiUrl = apiPreUrl+"vgood/marketplace/show";
		
		HeaderParams header = new HeaderParams();
		header.getKey().add("apikey");
		header.getValue().add(DeveloperConfiguration.apiKey);
		
		if(latitude != null && longitude != null){
			header.getKey().add("latitude");
			header.getValue().add(latitude);
			header.getKey().add("longitude");
			header.getValue().add(longitude);	
		}
		
		if(rows > 0){
			header.getKey().add("rows");
			header.getValue().add(Integer.toString(rows));
		}
		
		BeintooConnection conn = new BeintooConnection();
		String json = conn.httpRequest(apiUrl, header, null);
		Gson gson = new Gson();
		
		Vgood[] goods = gson.fromJson(json, Vgood[].class);
		
		return goods;	
	}
	
	public Vgood buyVgood(String vgoodExt, String userExt, boolean featured){
		String apiUrl = "";
		if(featured)
			apiUrl = apiPreUrl+"vgood/marketplace/buy/"+vgoodExt+"/"+userExt;
		else
			apiUrl = apiPreUrl+"vgood/marketplace/featured/buy/"+vgoodExt+"/"+userExt;
		
		HeaderParams header = new HeaderParams();
		header.getKey().add("apikey");
		header.getValue().add(DeveloperConfiguration.apiKey);
		
		BeintooConnection conn = new BeintooConnection();
		String json = conn.httpRequest(apiUrl, header, null,true);
		Gson gson = new Gson();
		
		Vgood vgood = gson.fromJson(json, Vgood.class);
		
		return vgood;
	}
}

File: beintoosdklight/mobile/ios/Classes/SDKBeintoo_Beta/API/BeintooVgood.m
/*******************************************************************************
 * Copyright 2011 Beintoo
 * 
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *   http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 ******************************************************************************/

#import "BeintooVgood.h"
#import "Beintoo.h"

@implementation BeintooVgood

@synthesize delegate,parser,generatedVGood,callingDelegate,_player;

-(id)init {
	if (self = [super init])
	{	
        parser = [[Parser alloc] init];
		parser.delegate = self;
		rest_resource = [[NSString alloc] initWithString:[NSString stringWithFormat:@"%@/vgood/",[Beintoo getRestBaseUrl]]];

		_player = [[BeintooPlayer alloc] init];
	}
    return self;
}

- (NSString *)restResource{
	return rest_resource;
}

// -------------------------------------------------------------------------------------
// Get SINGLE vgood, no delegate. The response will be sent to the main delegate
// -------------------------------------------------------------------------------------
+ (void)getSingleVirtualGood{
	
	NSString *guid = [Beintoo getPlayerID];
	CLLocation *loc	 = [Beintoo getUserLocation];
	
	if (guid == nil) {
		NSLog(@"Beintoo: unable to generate a vgood. No user logged.");
		return;
	}
	
	NSString *res;
	BeintooVgood *vgoodService = [Beintoo beintooVgoodService];

	if (loc == nil || (loc.coordinate.latitude <= 0.01f && loc.coordinate.latitude >= -0.01f)) {
		res = [NSString stringWithFormat:@"%@byguid/%@?allowBanner=true&rows=1",[vgoodService restResource],guid];
	}
	else
		res	= [NSString stringWithFormat:@"%@byguid/%@?latitude=%f&longitude=%f&radius=%f&allowBanner=true&rows=1",
				[vgoodService restResource],guid,loc.coordinate.latitude,loc.coordinate.longitude,loc.horizontalAccuracy];
	
	NSDictionary *params = [NSDictionary dictionaryWithObjectsAndKeys:[Beintoo getApiKey], @"apikey", nil];
	[vgoodService.parser parsePageAtUrl:res withHeaders:params fromCaller:VGOOD_SINGLE_CALLER_ID];
	NSLog(@"singleNODeleg %@",res);
}

// -------------------------------------------------------------------------------------
// Get SINGLE vgood, with delegate. The response will be sent to a custom delegate
// -------------------------------------------------------------------------------------
+ (void)getSingleVirtualGoodWithDelegate:(id)_delegate{
	NSString *guid = [Beintoo getPlayerID];
	CLLocation *loc	 = [Beintoo getUserLocation];
	

	if (guid == nil) {
		NSLog(@"Beintoo: unable to generate a vgood. No user logged.");
		return;
	}
	
	NSString *res;
	BeintooVgood *vgoodService = [Beintoo beintooVgoodService];

	if (loc == nil || (loc.coordinate.latitude <= 0.01f && loc.coordinate.latitude >= -0.01f)) {
		res = [NSString stringWithFormat:@"%@byguid/%@?allowBanner=true&rows=1",[vgoodService restResource],guid];
	}
	else
		res	= [NSString stringWithFormat:@"%@byguid/%@?latitude=%f&longitude=%f&radius=%f&allowBanner=false&rows=1",
			   [vgoodService restResource],guid,loc.coordinate.latitude,loc.coordinate.longitude,loc.horizontalAccuracy];
	

	vgoodService.callingDelegate = _delegate;
	NSDictionary *params = [NSDictionary dictionaryWithObjectsAndKeys:[Beintoo getApiKey], @"apikey", nil];
	[vgoodService.parser parsePageAtUrl:res withHeaders:params fromCaller:VGOOD_SINGLEwDELEG_CALLER_ID];
}

// -------------------------------------------------------------------------------------
// Get MULTIPLE vgood, no delegate. The response will be sent to the main delegate
// -------------------------------------------------------------------------------------
+ (void)getMultipleVirtualGood{
	NSString *guid = [Beintoo getPlayerID];	
	CLLocation *loc	 = [Beintoo getUserLocation];
		
	if (guid == nil) {
		NSLog(@"Beintoo: unable to generate a vgood. No user logged.");
		return;
	}
	NSString *res;
	BeintooVgood *vgoodService = [Beintoo beintooVgoodService];
	
	if (loc == nil || (loc.coordinate.latitude <= 0.01f && loc.coordinate.latitude >= -0.01f) 
				   || (loc.coordinate.longitude <= 0.01f && loc.coordinate.longitude >= -0.01f)) {
		res = [NSString stringWithFormat:@"%@byguid/%@?allowBanner=true&rows=3",[vgoodService restResource],guid];
	}
	else
		res	= [NSString stringWithFormat:@"%@byguid/%@?latitude=%f&longitude=%f&radius=%f&allowBanner=true&rows=3",
			   [vgoodService restResource],guid,loc.coordinate.latitude,loc.coordinate.longitude,loc.horizontalAccuracy];

	NSDictionary *params = [NSDictionary dictionaryWithObjectsAndKeys:[Beintoo getApiKey], @"apikey", nil];
	[vgoodService.parser parsePageAtUrl:res withHeaders:params fromCaller:VGOOD_MULTIPLE_CALLER_ID];
	NSLog(@"multipleNODeleg %@ and params %@",res,params);
}

// -------------------------------------------------------------------------------------
// Get MULTIPLE vgood, with delegate. The response will be sent to a custom delegate
// -------------------------------------------------------------------------------------
+ (void)getMultipleVirtualGoodWithDelegate:(id)_delegate{
	NSString *guid = [Beintoo getPlayerID];	
	CLLocation *loc	 = [Beintoo getUserLocation];
	
	if (guid == nil) {
		NSLog(@"Beintoo: unable to generate a vgood. No user logged.");
		return;
	}
	NSString *res;
	BeintooVgood *vgoodService = [Beintoo beintooVgoodService];
	
	if (loc == nil || (loc.coordinate.latitude <= 0.01f && loc.coordinate.latitude >= -0.01f)) {
		res = [NSString stringWithFormat:@"%@byguid/%@?allowBanner=true&rows=3",[vgoodService restResource],guid];
	}
	else
		res	= [NSString stringWithFormat:@"%@byguid/%@?latitude=%f&longitude=%f&radius=%f&allowBanner=true&rows=3",
			   [vgoodService restResource],guid,loc.coordinate.latitude,loc.coordinate.longitude,loc.horizontalAccuracy];
	
	vgoodService.callingDelegate = _delegate;
	NSDictionary *params = [NSDictionary dictionaryWithObjectsAndKeys:[Beintoo getApiKey], @"apikey", nil];
	[vgoodService.parser parsePageAtUrl:res withHeaders:params fromCaller:VGOOD_MULTIPLEwDELEG_CALLER_ID];
	NSLog(@"multipleDeleg %@",res);
}


+ (void)notifyVGoodGenerationOnUserDelegate{
	BeintooVgood *vgoodService = [Beintoo beintooVgoodService];
	id _callingDelegate = vgoodService.callingDelegate;
	
	if ([_callingDelegate respondsToSelector:@selector(didBeintooGenerateAVirtualGood:)]) {
		[_callingDelegate didBeintooGenerateAVirtualGood:[Beintoo getLastGeneratedVGood]];
	}	
}

+ (void)notifyVGoodGenerationErrorOnUserDelegate:(NSString *)_error{
	BeintooVgood *vgoodService = [Beintoo beintooVgoodService];
	id _callingDelegate = vgoodService.callingDelegate;

	if ([_callingDelegate respondsToSelector:@selector(didBeintooFailToGenerateAVirtualGoodWithError:)]) {
		[_callingDelegate didBeintooFailToGenerateAVirtualGoodWithError:_error];
	}	
}


#pragma mark -
#pragma mark API


- (void)showGoodsByUserForState:(int)state{
	NSString *res;
	if (state == TO_BE_CONVERTED) 
		res  = [NSString stringWithFormat:@"%@show/byuser/%@",rest_resource,[Beintoo getUserID]];
	if (state == CONVERTED) {
		res = [NSString stringWithFormat:@"%@show/byuser/%@/?onlyConverted=true",rest_resource,[Beintoo getUserID]];
	}
	NSDictionary *params = [NSDictionary dictionaryWithObjectsAndKeys:[Beintoo getApiKey], @"apikey", nil];
	[parser parsePageAtUrl:res withHeaders:params fromCaller:VGOOD_SHOWBYUSER_CALLER_ID];
}

- (void)sendGoodWithID:(NSString *)good_id asGiftToUser:(NSString *)ext_id_to{
	NSString *res  = [NSString stringWithFormat:@"%@sendasgift/%@/%@/%@",rest_resource,good_id,[Beintoo getUserID],ext_id_to];
	NSDictionary *params = [NSDictionary dictionaryWithObjectsAndKeys:[Beintoo getApiKey], @"apikey", nil];
	[parser parsePageAtUrl:res withHeaders:params fromCaller:VGOOD_SENDGIFT_CALLER_ID];
}

- (void)acceptGoodWithId:(NSString *)good_id{
	NSString *res  = [NSString stringWithFormat:@"%@accept/%@/%@",rest_resource,good_id,[Beintoo getUserID]];
	NSDictionary *params = [NSDictionary dictionaryWithObjectsAndKeys:[Beintoo getApiKey], @"apikey", nil];
	[parser parsePageAtUrl:res withHeaders:params fromCaller:VGOOD_ACCEPT_CALLER_ID];
}

#pragma mark -
#pragma mark MARKETPLACE

- (void)sellVGood:(NSString *)vGood_Id{
	NSString *res  = [NSString stringWithFormat:@"%@/marketplace/sell/%@/%@",rest_resource,vGood_Id,[Beintoo getUserID]];
	NSDictionary *params = [NSDictionary dictionaryWithObjectsAndKeys:[Beintoo getApiKey], @"apikey", nil];
	[parser parsePageAtUrlWithPOST:res withHeaders:params fromCaller:MARKET_SELLVGOOD_CALLER_ID];
	NSLog(@"CALL %@ with params %@",res,params);
}	

- (void)showGoodsToBuy{
	NSString *res  = [NSString stringWithFormat:@"%@/marketplace/show/",rest_resource];
	NSDictionary *params = [NSDictionary dictionaryWithObjectsAndKeys:[Beintoo getApiKey], @"apikey", nil];
	[parser parsePageAtUrlWithPOST:res withHeaders:params fromCaller:MARKET_GOODSTOBUY_CALLER_ID];
	NSLog(@"CALL %@ with params %@",res,params);
}
- (void)showGoodsToBuyFeatured{
	NSString *res  = [NSString stringWithFormat:@"%@/marketplace/show/?featured=true",rest_resource];
	NSDictionary *params = [NSDictionary dictionaryWithObjectsAndKeys:[Beintoo getApiKey], @"apikey", nil];
	[parser parsePageAtUrlWithPOST:res withHeaders:params fromCaller:MARKET_GOODSTOBUY_CALLER_ID];
	NSLog(@"CALL %@ with params %@",res,params);
}
- (void)buyGoodFromUser:(NSString *)vGood_Id{
	NSString *res  = [NSString stringWithFormat:@"%@/marketplace/buy/%@/%@",rest_resource,vGood_Id,[Beintoo getUserID]];
	NSDictionary *params = [NSDictionary dictionaryWithObjectsAndKeys:[Beintoo getApiKey], @"apikey", nil];
	[parser parsePageAtUrlWithPOST:res withHeaders:params fromCaller:MARKET_BUYVGOOD_CALLER_ID];
	NSLog(@"CALL %@ with params %@",res,params);
}
- (void)buyGoodFeatured:(NSString *)vGood_Id{
	NSString *res  = [NSString stringWithFormat:@"%@/marketplace/featured/buy/%@/%@",rest_resource,vGood_Id,[Beintoo getUserID]];
	NSDictionary *params = [NSDictionary dictionaryWithObjectsAndKeys:[Beintoo getApiKey], @"apikey", nil];
	[parser parsePageAtUrlWithPOST:res withHeaders:params fromCaller:MARKET_BUYVGOOD_CALLER_ID];
	NSLog(@"CALL %@ with params %@",res,params);
}


#pragma mark -
#pragma mark parser delegate

- (void)didFinishToParsewithResult:(NSDictionary *)result forCaller:(NSInteger)callerID{	
	switch (callerID){
			
		case VGOOD_SINGLE_CALLER_ID:{  // -------------------- SINGLE NO DELEGATE
			@try {
				//BLOG(@"multiple vgood result %@",result);
				if ([result objectForKey:@"messageID"]) {
					// No vgood is generated. a notification is sent to the main delegate
					[Beintoo notifyVGoodGenerationErrorOnMainDelegate:[result objectForKey:@"message"]];
					return;
				}
				NSArray *vgoodList = [result objectForKey:@"vgoods"];

				if (vgoodList == nil || [vgoodList count]<1) {
					[Beintoo notifyVGoodGenerationErrorOnMainDelegate:@"Beintoo Vgood Generation: Connection error or no Vgood retrieved."];
					return;
				}
				
				self.generatedVGood = [vgoodList objectAtIndex:0];
										
				BVirtualGood *vgood = [[BVirtualGood alloc] init];
				[vgood setVgoodContent:self.generatedVGood];
				[vgood setTheGood:self.generatedVGood];
				[Beintoo setLastGeneratedVgood:[vgood retain]];
				[vgood release];
						
				[Beintoo notifyVGoodGenerationOnMainDelegate];
			}
			@catch (NSException * e) {
				[_player logException:[NSString stringWithFormat:@"STACK: %@\n\nException: %@",[NSThread callStackSymbols],e]];
			}
		}
			break;
			
		case VGOOD_SINGLEwDELEG_CALLER_ID:{ // -------------------- SINGLE WITH DELEGATE
			@try {				
				//BLOG(@"multiple vgood result %@",result);
				if ([result objectForKey:@"messageID"]) {
					// No vgood is generated. a notification is sent to the main delegate
					[BeintooVgood notifyVGoodGenerationErrorOnUserDelegate:[result objectForKey:@"message"]];
					return;
				}
				NSArray *vgoodList = [result objectForKey:@"vgoods"];
				
				if (vgoodList == nil || [vgoodList count]<1) {
					[BeintooVgood notifyVGoodGenerationErrorOnUserDelegate:@"Beintoo Vgood Generation: Connection error or no Vgood retrieved."];
					return;
				}
				
				self.generatedVGood = [vgoodList objectAtIndex:0];
								
				BVirtualGood *vgood = [[BVirtualGood alloc] init];
				[vgood setVgoodContent:self.generatedVGood];
				[vgood setTheGood:self.generatedVGood];
				[Beintoo setLastGeneratedVgood:[vgood retain]];
				[vgood release];
				
				[BeintooVgood notifyVGoodGenerationOnUserDelegate];
			}
			@catch (NSException * e) {
				[_player logException:[NSString stringWithFormat:@"STACK: %@\n\nException: %@",[NSThread callStackSymbols],e]];
			}
		}
			break;
		
		case VGOOD_MULTIPLE_CALLER_ID:{  // -------------------- MULTIPLE NO DELEGATE
			@try {
				//BLOG(@"multiple vgood result %@",result);
				if ([result objectForKey:@"messageID"]) {
					// No vgood is generated. a notification is sent to the main delegate
					[Beintoo notifyVGoodGenerationErrorOnMainDelegate:[result objectForKey:@"message"]];
					return;
				}
				NSArray *vgoodList = [result objectForKey:@"vgoods"];
				
				if (vgoodList == nil || [vgoodList count]<1) {
					[Beintoo notifyVGoodGenerationErrorOnMainDelegate:@"Beintoo Vgood Generation: Connection error or no Vgood retrieved."];
					return;
				}

				if ([vgoodList count]==1 || ([[vgoodList objectAtIndex:0] objectForKey:@"isBanner"]!=nil) ) {
					// ------ We received only one vgood or a recommendation
					self.generatedVGood = [vgoodList objectAtIndex:0];
					
					BVirtualGood *vgood = [[BVirtualGood alloc] init];
					[vgood setVgoodContent:self.generatedVGood];
					[vgood setTheGood:self.generatedVGood];
					[Beintoo setLastGeneratedVgood:[vgood retain]];
					[vgood release];
					
					[Beintoo notifyVGoodGenerationOnMainDelegate];
				}
				if ([vgoodList count]>1 && ([[vgoodList objectAtIndex:0] objectForKey:@"isBanner"]==nil) ) { // ------ We received a list of vgood: this is a real multiple vgood
					
					BVirtualGood *vgood = [[BVirtualGood alloc] init];
					[vgood setTheGoodsList:vgoodList];
					[vgood setVgoodContent:[vgoodList objectAtIndex:0]];
					[vgood setIsMultiple:YES];
					[Beintoo setLastGeneratedVgood:[vgood retain]];
					[vgood release];
					
					[Beintoo notifyVGoodGenerationOnMainDelegate];
				}
			}
			@catch (NSException * e) {
				[_player logException:[NSString stringWithFormat:@"STACK: %@\n\nException: %@",[NSThread callStackSymbols],e]];
			}
		}
			break;
		
		case VGOOD_MULTIPLEwDELEG_CALLER_ID:{  // -------------------- MULTIPLE WITH DELEGATE
			@try {
				//BLOG(@"multiple vgood result %@",result);
				if ([result objectForKey:@"messageID"]) {
					// No vgood is generated. a notification is sent to the main delegate
					[BeintooVgood notifyVGoodGenerationErrorOnUserDelegate:[result objectForKey:@"message"]];
					return;
				}
				NSArray *vgoodList = [result objectForKey:@"vgoods"];
				
				if (vgoodList == nil || [vgoodList count]<1) {
					[BeintooVgood notifyVGoodGenerationErrorOnUserDelegate:@"Beintoo Vgood Generation: Connection error or no Vgood retrieved."];
					return;
				}
				
				if ([vgoodList count]==1 || ([[vgoodList objectAtIndex:0] objectForKey:@"isBanner"]!=nil) ) {
					// ------ We received only one vgood or a recommendation
					self.generatedVGood = [vgoodList objectAtIndex:0];
					
					BVirtualGood *vgood = [[BVirtualGood alloc] init];
					[vgood setVgoodContent:self.generatedVGood];
					[vgood setTheGood:self.generatedVGood];
					[Beintoo setLastGeneratedVgood:[vgood retain]];
					[vgood release];
					
					[BeintooVgood notifyVGoodGenerationOnUserDelegate];
				}
				if ([vgoodList count]>1 && ([[vgoodList objectAtIndex:0] objectForKey:@"isBanner"]==nil) ) { // ------ We received a list of vgood: this is a real multiple vgood
					
					BVirtualGood *vgood = [[BVirtualGood alloc] init];
					[vgood setTheGoodsList:vgoodList];
					[vgood setVgoodContent:[vgoodList objectAtIndex:0]];
					[vgood setIsMultiple:YES];
					[Beintoo setLastGeneratedVgood:[vgood retain]];
					[vgood release];
					
					[BeintooVgood notifyVGoodGenerationOnUserDelegate];
				}
			}	
			@catch (NSException * e) {
				[_player logException:[NSString stringWithFormat:@"STACK: %@\n\nException: %@",[NSThread callStackSymbols],e]];
			}
		}
			break;
			
				
		// ----------------------------------------------------------------------------------------------------
		// ----------------------------------------------------------------------------------------------------	
						
			
		case VGOOD_SHOWBYUSER_CALLER_ID:{
			[[self delegate]didGetAllvGoods:(NSArray *)result];
		}
			break;
			
		case VGOOD_ACCEPT_CALLER_ID:{
			//NSLog(@"ACCEPT RES: %@",result);
			if ([[self delegate] respondsToSelector:@selector(didAcceptVgood)]) {
				[[self delegate] didAcceptVgood];
			}
		}
			break;
			
		case VGOOD_SENDGIFT_CALLER_ID:{
			if ([[result objectForKey:@"message"] isEqualToString:@"OK"]) {
				[[self delegate]didSendVGoodAsGift:YES];
			}
			else {[[self delegate]didSendVGoodAsGift:NO];}
		}
			break;
			
			/* ------ MARKETPLACE ------- */
			
		case MARKET_SELLVGOOD_CALLER_ID:{
			//BLOG(@"sellVGOOD result: %@",result);
			if ([[result objectForKey:@"message"] isEqualToString:@"OK"]) {
				[[self delegate]didSellVGood:YES];
			}
			else {[[self delegate]didSellVGood:NO];}		
		}
			break;
		case MARKET_GOODSTOBUY_CALLER_ID:{
			//BLOG(@"vgood to buy result: %@",result);
			[[self delegate]didGetVGoodsToBuy:(NSArray *)result];
		}
			break;
		case MARKET_BUYVGOOD_CALLER_ID:{
			//BLOG(@"vgood bought result: %@",result);
			[[self delegate]didBuyVgoodWithResult:result];
		}
			break;
			
		default:{
			//statements
		}
			break;
	}	
}

- (void)dealloc {
	[parser release];
	[_player release];
	[rest_resource release];
	[super dealloc];
}


@end
